### PR TITLE
enhanced remove tag command, built test for remove tag command ,updated user guide

### PR DESCRIPTION
**Pull request by wenmogu**

### CODE REVIEW

File: docs/UserGuide.adoc [hunk 1]
```diff
@@ -241,6 +257,8 @@ e.g. `add n/James Ho p/22224444 e/jamesho@example.com a/123, Clementi Rd, 123466
 e.g. `delete 3`
 * *Edit* : `edit INDEX [n/NAME] [p/PHONE_NUMBER] [e/EMAIL] [a/ADDRESS] [t/TAG]...` +
 e.g. `edit 2 n/James Lee e/jameslee@example.com`
+* *Remove* : `remove TAG` +
+e.g. `remove friend`
 * *Find* : `find KEYWORD [MORE_KEYWORDS]` +
```
Comment (joanneong) [inline docs/UserGuide.adoc]: Documentation updates are fine, but just a general comment - you might want to split the changes into smaller commits next time (just in case you want to `revert` etc.) and to keep track of the incremental changes that happened in this PR

File: src/main/java/seedu/address/commons/core/Messages.java [hunk 2]
```diff
@@ -7,7 +7,9 @@
 
     public static final String MESSAGE_UNKNOWN_COMMAND = "Unknown command";
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
+    public static final String MESSAGE_INVALID_INPUT = "Invalid input value!";
     public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The person index provided is invalid";
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d persons listed!";
+    public static final String MESSAGE_TAG_NOT_FOUND = "Tag is not found.";
```
Comment (joanneong) [inline src/main/java/seedu/address/commons/core/Messages.java]: Just a thought: not sure if you want to remove this message because: (i) you can simply treat such cases as the tags having been removed anyway (e.g. for Zhen Yong's `clearhistory` command we will also get a success message even if no history was actually cleared, i.e. `clearhistory` is the first command) and (ii) this message is only used for the tag command, and is not a message that is likely to be shared across different commands (i.e. does not quite belong in `/commons`

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Remove a tag from the tag lists of the address book and all persons in the address book
+ */
+public class RemoveTagCommand extends UndoableCommand {
+    public static final String COMMAND_WORD = "remove";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ":Removes the tag identified by the user input from the tag lists of the address book and all"
+            + " persons in the address book.\n"
+            + "Parameters: TAG NAME (must be alphanumeric)\n"
```
Comment (joanneong) [inline src/main/java/seedu/address/logic/commands/RemoveTagCommand.java]: `TAGNAME` should be a single word to indicate that a single parameter suffices

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            model.removeTag(targetTag);
+        } catch (IllegalValueException ive) {
+            throw new CommandException(MESSAGE_INVALID_INPUT);
+        } catch (TagNotFoundException tnfe) {
+            throw new CommandException(MESSAGE_TAG_NOT_FOUND);
+        } catch (PersonNotFoundException pnfe) {
+            assert false : "The person to be updated is not found.";
+        }
+
+
```
Comment (joanneong) [inline src/main/java/seedu/address/logic/commands/RemoveTagCommand.java]: Unnecessary spacings here and in line 31/32 but minor issue

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public CommandResult executeUndoableCommand() throws CommandException {
+        try {
+            model.removeTag(targetTag);
+        } catch (IllegalValueException ive) {
+            throw new CommandException(MESSAGE_INVALID_INPUT);
+        } catch (TagNotFoundException tnfe) {
+            throw new CommandException(MESSAGE_TAG_NOT_FOUND);
+        } catch (PersonNotFoundException pnfe) {
```
Comment (joanneong) [inline src/main/java/seedu/address/logic/commands/RemoveTagCommand.java]: What is the `PersonNotFoundException` here for?

File: src/main/java/seedu/address/logic/parser/RemoveTagCommandParser.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import seedu.address.model.tag.Tag;
+
+/**
+ * Parses input argument and creates a new RemoveTagCommand object.
+ */
+public class RemoveTagCommandParser implements Parser<RemoveTagCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the RemoveTagCommand
+     * and returns a RemoveTagCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (joanneong) [inline src/main/java/seedu/address/logic/parser/RemoveTagCommandParser.java]: "does not conform **to** the expected format"

File: src/main/java/seedu/address/logic/parser/RemoveTagCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    public RemoveTagCommand parse(String args) throws ParseException {
+        try {
+            String trimmedArgs = args.trim();
+            System.out.println(Tag.isValidTagName(trimmedArgs));
```
Comment (joanneong) [inline src/main/java/seedu/address/logic/parser/RemoveTagCommandParser.java]: Hmm...this `println` statement...is not meant to be submitted right...?

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 8]
```diff
@@ -27,14 +27,24 @@
...
     public static final String VALID_ADDRESS_AMY = "Block 312, Amy Street 1";
     public static final String VALID_ADDRESS_BOB = "Block 123, Bobby Street 3";
+    public static final String VALID_ADDRESS_CEDAR = "Block 123, Cedary Street 3";
+    public static final String VALID_ADDRESS_DELON = "Block 123, Dragon Street 3";
     public static final String VALID_TAG_HUSBAND = "husband";
     public static final String VALID_TAG_FRIEND = "friend";
+    public static final String VALID_TAG_MUSHROOM = "mushroom";
+    public static final String VALID_TAG_FATTIE = "fattie";
 
     public static final String NAME_DESC_AMY = " " + PREFIX_NAME + VALID_NAME_AMY;
```
Comment (joanneong) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Very curious about why you added `CEDAR` and `DELON` into this part and `TypicalPersons` below since you did not seem to have utilised them for any testing at all in the end, so actually what was already given would have been sufficient...?

File: src/test/java/seedu/address/logic/commands/RemoveTagCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        // null -> returns false
+        assertFalse(removeFirstCommand.equals(null));
+
+        // different person -> returns false
+        assertFalse(removeFirstCommand.equals(removeSecondCommand));
+    }
+
+    /**
+     * Returns a {@code DeleteCommand} with the parameter {@code index}.
+     */
```
Comment (joanneong) [inline src/test/java/seedu/address/logic/commands/RemoveTagCommandTest.java]: Update the copied Javadoc comment

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 10]
```diff
@@ -19,6 +29,9 @@
 import seedu.address.model.person.ReadOnlyPerson;
 import seedu.address.model.person.exceptions.DuplicatePersonException;
 
+
+
```
Comment (joanneong) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Why did you add these spacings? Why did you add on to this file? (see above for the same comment)

**Review by joanneong — COMMENTED** [review]
The command seems to be working well; I tested it on my local machine and it works fine. However, I gave some comments on the code itself, so do take a look when you can

File: docs/UserGuide.adoc [hunk 11]
```diff
@@ -132,6 +132,22 @@ Deletes the 2nd person in the address book.
 `delete 1` +
 Deletes the 1st person in the results of the `find` command.
 
+=== Removing a tag: `remove`
```
Comment (Xenonym) [inline docs/UserGuide.adoc]: Could we have a more specific command name? A user might reasonably expect a `remove` command to function the same as `delete` (they mean the same thing after all). Consider `deletetag` - which also preserves consistency in the sense that all operations that delete stuff start with `delete`.

File: docs/UserGuide.adoc [hunk 12]
```diff
@@ -132,6 +132,22 @@ Deletes the 2nd person in the address book.
...
+
+Removes the specific tag from the address book. +
+Format: `remove TAG`
+
+****
+* Removes the tag with the specific name `TAG`.
+* The name refers to the name of tag, or more specifically, the words in the colored rectangles under the person's name.
```
Comment (Xenonym) [inline docs/UserGuide.adoc]: Reference to name unclear. I suggest just saying "Removes the tag `TAG`."
The "words in coloured rectangles" part seems oddly specific; presumably the user knows what tags are after adding them and seeing them show up in the UI,

File: src/main/java/seedu/address/logic/parser/RemoveTagCommandParser.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+            if (Tag.isValidTagName(trimmedArgs)) {
+                return new RemoveTagCommand(trimmedArgs);
+            } else {
+                throw new IllegalValueException("Invalid input value for tag!");
```
Comment (Xenonym) [inline src/main/java/seedu/address/logic/parser/RemoveTagCommandParser.java]: Why throw an `IllegalValueException` just to catch it immediately after? You can just throw the `ParseException` here directly.

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    public RemoveTagCommand(String targetTag) {
+        this.targetTag = targetTag;
+    }
+
+
+    @Override
+    public CommandResult executeUndoableCommand() throws CommandException {
+        try {
+            model.removeTag(targetTag);
+        } catch (IllegalValueException ive) {
```
Comment (Xenonym) [inline src/main/java/seedu/address/logic/commands/RemoveTagCommand.java:36]: If `removeTagCommandParser` accounts for illegal tag values, this should be an assertion rather than a new exception (if the parser checks the value before creating a `removeTagCommand`, it should be impossible for a `removeTagCommand` to have an illegal value).

**Review by Xenonym — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/model/AddressBook.java [hunk 15]
```diff
@@ -171,10 +171,13 @@ public void addTag(Tag t) throws UniqueTagList.DuplicateTagException {
      * @throws TagNotFoundException
      * @throws IllegalValueException
      */
-    public void removeTag(String tagGettingRemoved) throws TagNotFoundException, IllegalValueException {
+    public void removeTag(String tagGettingRemoved) throws TagNotFoundException,
```
Comment (Xenonym) [inline src/main/java/seedu/address/model/AddressBook.java]: How would this method throw a `PersonNotFoundException`?

File: src/main/java/seedu/address/model/AddressBook.java [hunk 16]
```diff
@@ -171,10 +171,13 @@ public void addTag(Tag t) throws UniqueTagList.DuplicateTagException {
...
+            IllegalValueException, PersonNotFoundException {
         Tag tagToRemove = tags.removeTag(tagGettingRemoved);
         for (Person person: persons) {
+            Person temp = person;
```
Comment (Xenonym) [inline src/main/java/seedu/address/model/AddressBook.java]: If you are directly accessing the `persons` list in the address book, `removeTag` should affect the underlying list directly - no need for the temp variable.

File: src/main/java/seedu/address/commons/core/Messages.java (hunk 2 above)
Comment (wenmogu) [reply src/main/java/seedu/address/commons/core/Messages.java]: i) emm what i imagined is the scenario that user wants to quickly find if a tag exists by using the removeTag command, since the user can redo this command, it's actually equivalent to findTag. Moreover, if the tags are sparsely distributed in the list, e.g. only one tag at the end of list which is super down the bottom, you might need multiple tries to delete the tag. therefore, i think reminding the user that certain tags do not exist is necessary.
ii) emm i m thinking of moving it to RemoveTagCommand class as inside the RemoveTagCommand class there is also a success message.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 16 above)
Comment (wenmogu) [reply src/main/java/seedu/address/model/AddressBook.java]: yup it affects the underlying list directly, but the changes are not immediately reflected on the console.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 15 above)
Comment (wenmogu) [reply src/main/java/seedu/address/model/AddressBook.java]: This is because of the `persons.setPerson` method used. This exception will not be thrown but if i dont write it here the project wont build.

File: src/main/java/seedu/address/commons/core/Messages.java (hunk 2 above)
Comment (joanneong) [reply src/main/java/seedu/address/commons/core/Messages.java]: (i) Well, since you find that it resembles a `find tag` command, I agree with you. Consequently, since I have already implemented a separate `find tag` command, do you think that that renders this redundant?
But I guess moving inside `RemoveTagCommand` class is fine, so go ahead

File: src/test/java/seedu/address/logic/commands/RemoveTagCommandTest.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    public void equals() {
+        RemoveTagCommand removeFirstCommand = new RemoveTagCommand(VALID_TAG_FRIEND);
+        RemoveTagCommand removeSecondCommand = new RemoveTagCommand(VALID_TAG_HUSBAND);
+
+        // same object -> returns true
+        assertTrue(removeFirstCommand.equals(removeFirstCommand));
+
+        // same values -> returns true
+        RemoveTagCommand removeFirstCommandCopy = new RemoveTagCommand(VALID_TAG_FRIEND);
+        assertTrue(removeFirstCommand.equals(removeFirstCommand));
```
Comment (joanneong) [inline src/test/java/seedu/address/logic/commands/RemoveTagCommandTest.java]: Incorrect test object used in line 66.

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java (hunk 14 above)
Comment (wenmogu) [reply src/main/java/seedu/address/logic/commands/RemoveTagCommand.java:36]: This is true; however in the RemoveTagCommandTest if I want to test if the RemoveTagCommand can handle the invalid input, there is no way for me to test for parser there, i.e. the RemoveTagCommand has to handle the exception itself but not an assertion...I think if there is another command like `assertFalse` that allows me to test for assertion then i will change this part to assertion; or else i just remove the test? or I write a removeTagCommandParserTest?

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java (hunk 5 above)
Comment (wenmogu) [reply src/main/java/seedu/address/logic/commands/RemoveTagCommand.java]: this is for the `persons.setPerson` used in AddressBook.java, which allows for instand update of the list displayed.

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 18]
```diff
@@ -27,14 +27,24 @@
...
     public static final String VALID_EMAIL_AMY = "amy@example.com";
     public static final String VALID_EMAIL_BOB = "bob@example.com";
+    public static final String VALID_EMAIL_CEDAR = "cedar@example.com";
+    public static final String VALID_EMAIL_DELON = "delon@example.com";
     public static final String VALID_ADDRESS_AMY = "Block 312, Amy Street 1";
     public static final String VALID_ADDRESS_BOB = "Block 123, Bobby Street 3";
+    public static final String VALID_ADDRESS_CEDAR = "Block 123, Cedary Street 3";
+    public static final String VALID_ADDRESS_DELON = "Block 123, Dragon Street 3";
     public static final String VALID_TAG_HUSBAND = "husband";
     public static final String VALID_TAG_FRIEND = "friend";
```
Comment (joanneong) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Update: I have taken a closer look at the tests, and have concluded that *`Cedar` and `Delon` are still unfortunately redundant* -- please remove them since they add no value to the tests :(

Explanation:
Observe that you can simply add `Amy` and `Bob` into the list returned by `getTypicalPersons()` method in `TypicalPersons`and all your tests will still pass. The reason why they did not pass at first without `Amy`, `Bob`, `Cedar` and `Delon` is just because `VALID_TAG_FRIEND` is stated to be `friend`, not `friends`, so it throws an exception claiming that the tag does not exist in the original list given by the original `getTypicalPersons()` method.

**Review by joanneong — CHANGES_REQUESTED** [review]

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java (hunk 18 above)
Comment (joanneong) [reply src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Update 2: Upon more meticulous testing (this time by running `./gradlew allTests` and waiting some 22 minutes), I propose a superior alternative to resolving this issue: replace `VALID_TAG_FRIEND` with `friends` in `RemoveTagCommandTest.java`. This will prevent the need to edit `TypicalPersons` list, while still ensuring that the tests are all carried out properly. Alternatively, you can create a `VALID_TAG_FRIENDS` with the value `friends` in `CommandTestUtil` instead

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java (hunk 14 above)
Comment (Xenonym) [reply src/main/java/seedu/address/logic/commands/RemoveTagCommand.java:36]: It is okay to remove that particular test I think as `RemoveTagCommand` must have a valid tag as a precondition, so it should never be encountered anyway.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 15 above)
Comment (Xenonym) [reply src/main/java/seedu/address/model/AddressBook.java]: If it is reasonable to assume that this method will only deal with persons that exist, I suggest catching the `PersonNotFoundException` specifically and cause an assertion to indicate that it is an "impossible" condition.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 16 above)
Comment (Xenonym) [reply src/main/java/seedu/address/model/AddressBook.java]: Noted!

File: src/main/java/seedu/address/logic/commands/RemoveTagCommand.java (hunk 14 above)
Comment (wenmogu) [reply src/main/java/seedu/address/logic/commands/RemoveTagCommand.java:36]: okie then i will write a RemoveTagCommandParserTest to cover this

**Review by Xenonym — APPROVED** [review]
